config: refactor datatype configure

hzhou opened the pull request:
## Pull Request Description

`configure` currently handles both feature detection, type matching, arithmetic, as well as embedding the bit masks and sequence number of the built-in data types. This PR tries to let configure only do feature detection, as it was designed to do, and move the type matching and arithmetic to a Perl script, and leave the HANDLE logic remain in `mpi.h`.

This PR is a overhaul started with PR #3811, #3812, and #3814. Some of commits may go back to these PR, or this PR may out date them.

## Expected Performance Changes

Reduce the complexity in configure and it should run a bit faster as well.

## Known Issues

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
  Issues related to cross file: 
       #393 -- lot of history, but what is the final word?
       #935  -- no longer true, close.
       #3753 -- asking for a script to generate cross file; this PR fixes it.
  Issues related to datatypes:
       #3289 -- AC_TYPE_INT8_T with pgi compiler
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [x] Add Devel Docs in the `doc/` directory for any new code design
* [x] Commits are self-contained and do not do two things at once
* [x] Passes tests (included warning check)

hzhou commented:
test:mpich/custom
testlist: attr
testlist: cxx
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:mpich/custom
testlist: attr
testlist: cxx
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:mpich/custom
testlist: attr
testlist: cxx
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:mpich/custom
testlist: attr
testlist: cxx
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:mpich/custom
testlist: attr
testlist: coll/allred 4
testlist: cxx
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:mpich/custom
testlist: attr
testlist: coll/allred 4
testlist: cxx
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:mpich/custom
compiler=intel
testlist: pt2pt
testlist: coll/allred 4
testlist: cxx
testlist: f77
testlist: f90
testlist: f08

hzhou commented on confdb/aclocal_f77.m4:
```diff
@@ -800,17 +824,17 @@ AC_LANG_PUSH([C])
     AC_LANG_PROGRAM([],[long long a;])
 ],[
-    AC_DEFINE(HAVE_LONG_LONG,1,[Define if long long allowed])
-])
```
User/device may pass-in options to disable "long long". All `HAVE_type` macros are defined by `config_filter`; `configure.ac` only need check `sizeof`.

hzhou commented on maint/config_filter.pl:
```diff
@@ -0,0 +1,332 @@
+print Out @lines;
+close Out;
+open Out, ">src/include/mpichconf.h" or die "Can't write src/include/mpichconf.h.\n";
+foreach my $l (@mpichconf_lines){
```
Modifying `src/include/mpichconf.h` is a convenience. A better solution may be to create a separate header file and included in `mpi_ext.h` (as ROMIO probably need it too).

hzhou commented:
test:mpich/custom
compiler=intel
testlist: pt2pt
testlist: coll/allred 4
testlist: cxx
testlist: f77
testlist: f90
testlist: f08

hzhou commented on configure.ac:
```diff
@@ -1635,26 +1631,12 @@ if test -n "$perform_dbgmutex" ; then
-	      sed -e 's/^/export /g' -e 's/=.*//g' > confcross
-	. confcross
-	rm -f confcross
     fi
```
To coordinate between sub-configure is just too messy. I suggest a solution is to have a clean output from MPICH's configure --`mpichconf.h` or even `mpich_config.sh`  -- that other modules can parse or source. That way, we'll have a consolidated place to coordinate between packages.

For datatypes, currently we set unavailable datatypes to `MPI_DATATYPE_NULL`. Can we instead just expose `HAVE_type` macros in `mpi.h`? Either is non-standard interface, but if we are going to use it -- dtpool, romio -- we might as well use a straight-forward mechanism. We might as well expose the datatype sizes and c type conversions too.

raffenet replied:
`HAVE_type` would not be portable between different MPI libraries.

hzhou commented:
test:mpich/custom
compiler=intel
testlist: pt2pt/sendrecv1 2 arg=-type=MPI_INT:4+MPI_DOUBLE:8 arg=-sendcnt=1 arg=-recvcnt=1 arg=-seed=37 arg=-testsize=1024
testlist: coll/allred 4
testlist: cxx
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:mpich/custom
compiler=intel
testlist: pt2pt/sendrecv1 2 arg=-type=MPI_INT:4+MPI_DOUBLE:8 arg=-sendcnt=1 arg=-recvcnt=1 arg=-seed=37 arg=-testsize=1024
testlist: coll/allred 4
testlist: cxx
testlist: f77
testlist: f90
testlist: f08

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
test:jenkins/ch3/most 
test:jenkins/ch4/most
--

hzhou commented:
@pavanbalaji @raffenet @gcongiu 
This PR is nearly ready and it will need some of your review and comments to get to its final shapes. Right now it passes my set of preliminary tests, and it is waiting for the comprehensive tests. This PR trims more than 1500 lines off of `configure.ac` and a big step toward making it cleaner. I tried to separate the some of the commits, but the main commit is so connected and remains to be a big one. Feel free to point out and asking me to split further (into more commits or separate PR).

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most
test:mpich/custom

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most
test:mpich/custom

hzhou commented:
test:mpich/custom/centos32

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most
test:mpich/custom/centos32

pavanbalaji commented on confdb/aclocal_cc.m4:
```diff
@@ -696,6 +696,8 @@ dnl	eight
 dnl In addition, a "Could not determine alignment" and a "error!"
 dnl return is possible.  
+dnl 
+dnl Note: this macro assumes AC_CHECK_SIZEOF([long long]) has been run
```
We can run `AC_CHECK_SIZEOF([long long])` again inside this macro, instead of assuming that it has already been run.

pavanbalaji replied:
Btw, this is not a refactoring patch.  Refactoring generally refers to moving existing code with no changes to algorithmic logic.  Changing `#ifdef HAVE_LONG_LONG_INT` to `#if SIZEOF_LONG_LONG > 0` is a change in algorithmic logic, and hence not refactoring.  The comment message needs to be changed.

hzhou replied:
In the current code, probably for legacy reasons, we allow disabling the support of `long long` or `long double`. And for the same reasons that user may disable the `long long/long double`, they may wish to specify the size -- for example, cross-compilation where host system disagree with build system. To support that, we need a central point to do the size checking. By setting cache values, having multiple `AC_CHECK_SIZEOF` may still work. However, I thought to have a logical stage in configure where we check compiler, then check basic data type sizes, and then alignment, will make the structure of `configure` easier to work with.

I agree it is beyond a typical refactoring for this PR. I'll revise the descriptions.

pavanbalaji replied:
Note that we started requiring C99 quite recently.  So, all the additional checks for C99 types should be removed (or an abort added if those are not available).

hzhou replied:
Ok.

pavanbalaji commented on confdb/aclocal_cc.m4:
```diff
@@ -714,7 +716,7 @@ int main( int argc, char *argv[] )
     struct { char a; int b; char c; } char_int_char;
     struct { char a; short b; char c; } char_short_char;
-#ifdef HAVE_LONG_LONG_INT
+#if SIZEOF_LONG_LONG > 0
```
`long long` is C99.  At this point we assume C99 in mpich.  So if `sizeof(long long)` is `0`, we need to abort.

hzhou replied:
The current code does not assume the availability of "long long" or "long double". 

Whether to support "long long" or "long double" is more about using the library rather than the library itself. So even the library need C99 to build, the user may still prefer those types to be invalid -- I am not making a strong argument, just playing a cautious role here.

I am OK to require the availability of these types.

pavanbalaji replied:
I see no reason to retain those checks.  Those are C99, and we require C99 compilers at this point.

hzhou replied:
Ok.

pavanbalaji commented on confdb/aclocal_cc.m4:
```diff
@@ -827,6 +829,8 @@ dnl     sixteen
 dnl In addition, a "Could not determine alignment" and a "error!"
 dnl return is possible.  
+dnl 
+dnl Note: this macro assumes AC_CHECK_SIZEOF([long double]) has been run
```
We should rerun `AC_CHECK_SIZEOF([long double])` instead of assuming.

hzhou replied:
The intention is more about setting clear stages in `configure`. Otherwise, support of configure options and support cross-files become messy. Will it be more acceptable if we set stage-variable and assert the stage variable here rather than to assume in a comment?

pavanbalaji replied:
The assumption in the comment is not a good choice.  If you don't want to rerun sizeof, then we should at least assert.

hzhou replied:
Agreed.

pavanbalaji commented on confdb/aclocal_cc.m4:
```diff
@@ -845,7 +849,7 @@ int main( int argc, char *argv[] )
     struct { char a; double b; } char_double;
     struct { double b; char a; } double_char;
-#ifdef HAVE_LONG_DOUBLE
+#if SIZEOF_LONG_DOUBLE > 0
```
`long double` seems to have been there since C89.  Again, I recommend we abort if it's not present.

https://en.wikipedia.org/wiki/Long_double

hzhou replied:
Ok.

pavanbalaji commented:
I tried to review this PR, but gave up after some time.  The patches in this PR combine a lot of unrelated changes making it hard to review.  This PR needs to be cleaned up a lot before it's ready for review again.

hzhou commented:
> 
> 
> I tried to review this PR, but gave up after some time. The patches in this PR combine a lot of unrelated changes making it hard to review. This PR needs to be cleaned up a lot before it's ready for review again.

Thanks for start reviewing this PR. I agree the first commit is way too big. Can we do it parts by parts? The rest of the commits are trivial and added here to help the testing. Should I split them into separate PRs and get them reviewed first? After that, I can try to split the first commit into parts as well, but I would need the help of reviewers to figure out the best way to split.

pavanbalaji commented:
> > I tried to review this PR, but gave up after some time. The patches in this PR combine a lot of unrelated changes making it hard to review. This PR needs to be cleaned up a lot before it's ready for review again.
> 
> Thanks for start reviewing this PR. I agree the first commit is way too big. Can we do it parts by parts? The rest of the commits are trivial and added here to help the testing. Should I split them into separate PRs and get them reviewed first? After that, I can try to split the first commit into parts as well, but I would need the help of reviewers to figure out the best way to split.

You can move the smaller (unrelated) patches either to early commits in this PR or a separate PR.  Right now, it does too many different things.  We can't merge it in this state.

hzhou commented:
> You can move the smaller (unrelated) patches either to early commits in this PR or a separate PR. Right now, it does too many different things. We can't merge it in this state.

OK. I'll work on it tomorrow.

hzhou commented:
* rebased to master
* dropped a few unrelated commits
* split the original big commit into smaller units, hopefully, easier for review. 

@pavanbalaji Could you take a look and provide some feedback to see if I am on the right track? I will work on the tests (in case I have messed up the conflicts somewhere) tomorrow.

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
test:jenkins/ch4/most

hzhou commented:
Rebased to master.

hzhou commented:
test:mpich/warnings

hzhou commented on src/include/mpi.h.in:
```diff
@@ -102,23 +102,119 @@ extern "C" {
-#define MPI_LONG_LONG_INT  ((MPI_Datatype)@MPI_LONG_LONG@)
-#define MPI_UNSIGNED_LONG_LONG ((MPI_Datatype)@MPI_UNSIGNED_LONG_LONG@)
-#define MPI_LONG_LONG      MPI_LONG_LONG_INT
+#define MPI_CHAR           ((MPI_Datatype)0x4c000001)
```
That is a good suggestion.

pavanbalaji commented:
Large refactoring change.  Should be delayed till the next major release.
